### ISSUE TITLE
Multiple identical Message Events are not honoured with Publish Message

### ISSUE THREAD

**Issue by arpitagarwal78** [zeebe-io/zeebe#4644]
**Describe the bug**
Identical message events within parallel branch in workflow. On publishing of the message event only one parallel branch gets the message and the simultaneous branches ignore the message.

**To Reproduce**
![image](https://user-images.githubusercontent.com/37263155/83472166-3e989400-a4a4-11ea-985d-472349cf8be3.png)
1. A parallel gateway with two identical `StartMe` message.
2. On publishing `StartMe` message only one branch is honoured.

**Expected behaviour**
All the parallel events should be taken into considerations.

**Environment:**
 - OS: Mac
 - Zeebe Version: 0.21.0
 - Configuration: NA

PFA - 
[StartMe.bpmn.txt](https://github.com/zeebe-io/zeebe/files/4714379/StartMe.bpmn.txt)

**Comment by saig0** [comment]
@arpitagarwal78, thank you for reporting.

This behavior is expected and described in the documentation: https://docs.zeebe.io/reference/message-correlation/message-correlation.html#message-cardinality

> A message is correlated only once to a workflow (based on the BPMN process id), across all versions of this workflow. If multiple subscriptions for the same workflow are opened (by multiple workflow instances or within one instance) then the message is correlated only to one of the subscriptions.

Why do you want to correlate the same message twice within one workflow instance?

Please share your example so that we can find a solution :)

**Comment by arpitagarwal78** [comment]
Use case which we are targeting is executing multiple parallel sub workflow based on identical event.

In our case currently it lies as a boundary event which when send should also automatically start an internal sub workflow

Ex.

![BoundaryEvent](https://user-images.githubusercontent.com/37263155/83498472-97365400-a4d9-11ea-9175-5fd84791ab9c.png)

w.r.t figure 

1. On receiving of `BoundaryEvent` at any time should trigger some workflow currently not included within diagram 
2. This `BoundaryEvent` can we received while in loop or otherwise.
3. If in loop it should mark the completion of 1st parallel sequence and start the 2nd parallel sequence.


PFA - Sample.bpmn file
[Sample.bpmn.txt](https://github.com/zeebe-io/zeebe/files/4715890/Sample.bpmn.txt)



Similar to the same lines we also want a feature that Zeebe should interpolate all the identical events at same shot. Since we want to hook to simultaneously executing multiple workflows for similar event

**Comment by arpitagarwal78** [comment]
JFR @korthout 
https://github.com/zeebe-io/zeebe/issues/4645#issuecomment-637693197

On posting of `IntermediateThrowEvent`(BoundaryEvent) the internal subprocess workflow(`blah`) should stop executing and should never loop.


Just to validate that loop is the only issue I tried the same thing without loop

Ex.
<img width="1125" alt="Screenshot 2020-06-02 at 11 02 11 PM" src="https://user-images.githubusercontent.com/37263155/83550907-32a0e680-a525-11ea-93fd-85a8ccacd917.png">

still the workflow is stuck in `Active` state


Next thing i tried I disabled the `blah` worker and tried posting `BoundaryEvent` still the workflow was stuck in `Active` state

<img width="984" alt="Screenshot 2020-06-02 at 11 07 12 PM" src="https://user-images.githubusercontent.com/37263155/83551458-00dc4f80-a526-11ea-9f56-dd9c0f2e07a8.png">

**Comment by arpitagarwal78** [comment]
@saig0 
Please let me know if any update

**Comment by saig0** [comment]
> Next thing i tried I disabled the blah worker and tried posting BoundaryEvent still the workflow was stuck in Active state

The workflow instance is waiting in both cases at the parallel gateway. The gateway has 3 incoming sequence flows. In order to activate the gateway, all incoming sequence flows must be taken. 
Looking at the workflow, this is not intended because either the subprocess is completed or the boundary event is triggered. This can be solved by joining the two flows using an [exclusive gateway](https://docs.zeebe.io/bpmn-workflows/exclusive-gateways/exclusive-gateways.html) that is connected to the parallel gateway.

> 1. On receiving of BoundaryEvent at any time should trigger some workflow currently not included in diagram
> 2. This BoundaryEvent can be received in loop or otherwise.
> 3. If in loop it should mark the completion of 1st parallel sequence and start the 2nd parallel sequence.

Depending on your concrete workflow, it is maybe possible to combine the multiple message catch/boundary events. For example, using a message event subprocess or throwing another message after the boundary event/event subprocess is triggered.

I'm not sure what you mean with the 3 point. 

> On the same lines we also want a feature that Zeebe should interpolate all the identical events, this will help us executing multiple workflows simultaneously

This sounds like a feature request for a signal event. The idea is that a signal is broadcast to many receives.

**Comment by arpitagarwal78** [comment]
@saig0 
Thanks for the help

> On the same lines we also want a feature that Zeebe should interpolate all the identical events, this will help us executing multiple workflows simultaneously

Yes it would be great if this feature is included in workflow execution.

Please let me know How you feel for the same ?

**Comment by saig0** [comment]
@arpitagarwal78, I think that signal events can be useful in some situations. But it is also a bigger topic and some points of the behavior need to be declared (e.g. correlation via correlation-key, buffering, etc.). Feel free to create a new issue.

Please note that it might take some time until signal events are available (depending on the internal priorities and the input from the community). So, it could make sense to look for an alternative modeling in the meantime.

**Comment by arpitagarwal78** [comment]
@saig0 thanks a lot
opened a new issue
https://github.com/zeebe-io/zeebe/issues/4680

**Comment by saig0** [comment]
Closing this issue in favor of #4680
